app/models/ArrivalUpdate.scala

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2020 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package models
+
+import cats._
+import cats.data._
+import cats.implicits._
+import play.api.libs.json.JsObject
+import play.api.libs.json.Json
+import play.api.libs.json.Writes
+
+case class MessageStatusUpdate(messageId: MessageId, messageStatus: MessageStatus)
+
+object MessageStatusUpdate {
+  implicit def arrivalStateUpdate(implicit writes: Writes[MessageStatus]): ArrivalModifier[MessageStatusUpdate] =
+    ArrivalModifier(
+      value =>
+        Json.obj(
+          "$set" ->
+            Json.obj(
+              s"messages.${value.messageId.index}.status" -> value.messageStatus
+            )
+      )
+    )
+}
+
+case class ArrivalUpdate(arrivalUpdate: Option[ArrivalStatus], messageUpdate: Option[MessageStatusUpdate])
+
+object ArrivalUpdate {
+
+  implicit object ArrivalUpdateSemiGroupInst extends Semigroup[ArrivalUpdate] {
```

Comment:
Semigroup?

Reply:
I've started modelling all the updates that can happen as a `Semigroup`, which means that we can combine them together. So given two `ArrivalUpdate`, we can combine them together to get an `ArrivalUpdate` that (in our case) is a combination of the two. Using cats here means that I get a few extra things that we don't need to do manually like it being associative and we can use the cats-laws for some testing.

An example of how we can use this is given an ArrivalUpdate to the Arrival's state, and an ArrivalUpdate to a Message's state, we can combine them together to get an update type that captures both these operations, and then we can pass that along to be interpreted by repository layer. We don't leverage this yet, we still need to move those out the repository layer and get an intermediate layer to construct those ArrivalUpdates for us